### ISSUE TITLE
ResourceId matcher doesn't work correctly when primary has the same name and namespace but different kind

### ISSUE THREAD

**Issue by AleksanderBrzozowski** [operator-framework/java-operator-sdk#2368]
## Bug Report

In a `KubernetesDependent` class, we call a method `getSecondaryResource(primary, context)`.

By default, this method finds a secondary resource based on a primary `ResourceId`, which is a combination of name and namespace. The problem arises when there are two different secondary resources, which are using the same primary name and namespace, but the primary is a different resource, for example:

```yaml
# First secondary resource
  ownerReferences:
  - apiVersion: foo/v1
    kind: Foo
    name: my-resource
    uid: <uid>

# Second secondary resource
  ownerReferences:
  - apiVersion: bar/v1
    kind: Bar
    name: my-resource
    uid: <uid>
```

These two secondary resources have a different primary resource. However, the default implementation doesn't take into account resource `kind` and `apiVersion`.

#### What did you do?

When calling `getSecondaryResource(primary, context)`, we got an error:

```
IllegalStateException("More than 1 secondary resource related to primary")
```

even though there is only one secondary resource.

#### What did you expect to see?

One secondary resource is returned, matched by a primary.

#### What did you see instead? Under which circumstances?

We got an error: `More than 1 secondary resource related to primary`

#### Environment

The issue doesn't seem to be dependent on Kubernetes / Java version.

JOSDK version: 4.8.0

#### Possible Solution

This issue might be overcome by implementing primary to secondary mapper. 

However, in the documentation, there is this sentence:

> Seasoned Kubernetes users already know one way to track this parent-child kind of relationship: using owner references. Indeed, that’s how the SDK deals with this situation by default as well, that is, if your controller properly set owner references on your secondary resources, the SDK will be able to follow that reference back to your primary resource automatically without you having to worry about it.

Which is not true in this case 🙂 

Basically, I would like to know what are your thoughts on this, if we don't want to change the default implementation, then we need to make changes on our side 🙂

**Comment by csviri** [comment]
Hi @AleksanderBrzozowski , this is a known issue, will be properly fixed in v5 for dependent resources:
https://github.com/operator-framework/java-operator-sdk/issues/2349

We might want to create a patch for 4.8.x but,  until you can create you own secondary to primary mapper to take into account he kind. 

If no objections will close this issue as a duplicate.

**Comment by AleksanderBrzozowski** [comment]
Yeah, we can close it. Basically, we have a workaround for this, and it's better in our case to wait for a new release 🙂

**Comment by csviri** [comment]
Good to hear, thx for the issue, v5 might take some time tbh, but will be there!
